src/Workspaces/Core/Portable/Progress/CodeAnalysisProgress.cs

```diff
@@ -0,0 +1,149 @@
+
+    /// <summary>
+    /// Adds the requested number of incomplete items to the UI showing the progress of the current operation.  This is
+    /// commonly presented with a progress bar.  An optional <paramref name="description"/> can also be provided to
+    /// update the UI accordingly (see <see cref="Description"/>).
+    /// </summary>
+    /// <param name="count">The number of incomplete items left to perform.</param>
+    /// <param name="description">Optional description to update the UI to.</param>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.AddIncompleteItems(20));
+    /// </example>
+    public static CodeAnalysisProgress AddIncompleteItems(int count, string? description = null)
+        => new()
+        {
+            IncompleteItemsValue = count >= 0 ? count : throw new ArgumentOutOfRangeException(nameof(count)),
+            DescriptionValue = description,
+        };
+
+    /// <summary>
+    /// By default, indicates that an item of work has transitioned from being incomplete (see <see
+    /// cref="AddIncompleteItems"/> to complete.  This is commonly presented with a progress bar. An optional <paramref
+    /// name="description"/> can also be provided to update the UI accordingly (see <see cref="Description"/>).
+    /// </summary>
+    /// <remarks>
+    /// Multiple items of work can be transitioned to be complete by passing an explicit value to <paramref
+    /// name="count"/>.
+    /// </remarks>
+    /// <param name="count">The number of items that were completed.  Defaults to <c>1</c> for the trivial case where
+    /// only a single item has been completed.  Must be greater than or equal to 1.</param>
+    /// <param name="description">Optional description to update the UI to.</param>
+    /// <example>
+    /// progress.Report(CodeAnalysisProgress.CompleteItem());
+    /// </example>
+    public static CodeAnalysisProgress CompleteItem(int count = 1, string? description = null)
+        => new()
+        {
+            CompleteItemValue = count >= 1 ? count : throw new ArgumentOutOfRangeException(nameof(count)),
+            DescriptionValue = description,
+        };
+
+    /// <summary>
+    /// Indicates that all progress should be reset for the current operation. This is normally done when the code
+    /// action is performing some new phase and wishes for the UI progress bar to restart from the beginning.
+    /// </summary>
+    /// <remarks>
+    /// Currently internal as only roslyn needs this in the impl of our suggested action (we use a progress bar to
+    /// compute the work, then reset the progress to apply all the changes).  Could be exposed later to 3rd party code
+    /// if a demonstrable need is presented.
+    /// </remarks>
+    internal static CodeAnalysisProgress Clear()
+        => new() { ClearValue = true };
+}
+
+internal sealed class CodeAnalysisProgressTracker(Action<string?, int, int>? updateAction) : IProgress<CodeAnalysisProgress>
```

Comment:
Learning opportunity for me, when is it preferable to have multiple top level class in the same file in Roslyn?

Reply:
i often do it when there are things that are bound extremely tightly that exist for each other, which are very small.  in this case, there's just a type which exists as a forwarding/bridging type.  So i kept them together.  But you're always welcome to split apart, or use best judgement in aggregating.  We have no hard and fast rule. 

Another common case is:

```
interface IService
{
}

class DefaultServiceImpl : IService
{
}
```

Reply:
ok, thanks